### ISSUE TITLE
[question] Does netdata provide kind of issue impact analysis?

### ISSUE THREAD

**Issue by archenroot** [firehol/netdata#2874]
Hi, imagine I am in situation when component on one server fails and I also have network outage on some router, could by anywhere for example on the way from client A to the database depends on failed component. This is naive scenario introducing 2 failures - 1 hardware,  1- software (your are free to create your own). Now I am interested in failure impact or generate error or some message to pushed into message broker from netdata saying:

Part 1)
HW failure - network router Alfa
SW failure - component B 

Providing graph of affected systems:
A
B
C

maybe better in form of real graph visualization + text information. I can click on the affected system and can be provided with contact information.

Part 2)
We live in world of microservices in these days, so REST API and MB event driven integrations. Now depends on affected systems I might be required to either notify admin/operator/user:
Admin - repair it
Operator - contact someone to help you
User - there is something wrong and we work on this

Dropping message into message queue/topic (any kind of protocol supported - AMQP, JMS, etc.) or call REST API would be really handy. 

I don't want to suggest to implement everything here now, but maybe suggest possible ideas for future evolution path of netdata. I understand few of ideas mentioned here are already in space of issue resolution or ticket management, which possibly netdata won't aim to breach...

Any feedback on these ideas welcomed.

**Comment by ktsaou** [comment]
Nice.

The easy part 2: netdata offers many ways to do this:

1. you can use the `custom` notifications to hook whatever you like, in parallel with alarm notifications. 
2. you can change the `exec` line of alarms to execute your scripts instead of `alarm-notify.sh`.
3. `alarm-notify.sh` supports **roles**. You can assign alarms to multiple roles, each role can have multiple recipients across multiple channels and each recipient on each channel may receive only critical notifications. So, it is simple to say that alarm X will:

- send notifications to slack channel `sysadmin`
- send emails to ops@example.com and sysadmins@example.com
- send SMS notifications to NUMBER1 and only critical notifications to NUMBER2
- execute your command `/opt/company-scripts/restart-service.sh` (keep in mind it will be run as user `netdata`)
etc.

So, I think netdata already supports this. Also, `alarm-notify.sh` already supports kafka. I guess it would be quite simple to add more queues (PRs accepted).

---

Regarding the impact analysis, I need to think a bit about it.

Given that netdata runs always on a system, it could somehow "know" a few things.

For example, It could learn what is "normal" for this system. So, you could ask "what is abnormal now?" and netdata would give you a sorted list of things that are not so common.

I have tried to find some AI library to embed into netdata, so that it will continuously learn (without luck). 

But even simpler solutions could work: a simple hourly average per chart for each of the 24 hours of the day (baseline), could be used to compare the last-hour average and provide a sorted list of charts with the biggest delta compared to the baseline for this hour.

Or, it could maintain a histogram for each chart, and give you the charts that seem to be at the edge 5% of the histograms.

Or, maintain trend vectors for each chart, and give the charts that collect values outside their trend.

I think netdata should expand a bit upper (i.e. more high level features, like the ones above), but it should also go deeper (i.e. even higher detail of what is really happening).

**Comment by archenroot** [comment]
@ktsaou - first of all thank you for attention. Netdata is really great piece of software and I don't want to harm it with something which doesn't belong there. :D

> 1. you can use the custom notifications to hook whatever you like, in parallel with alarm notifications.
> 2. you can change the exec line of alarms to execute your scripts instead of alarm-notify.sh.
I am new to netdata, so in process of discovery. This is greate I can implement any kin of either REST client or message broker message submission
> 3. alarm-notify.sh supports roles. You can assign alarms to multiple roles, each role can have multiple recipients across multiple channels and each recipient on each channel may receive only critical notifications. So, it is simple to say that alarm X will:
> 4. send notifications to slack channel sysadmin
> 5. send emails to ops@example.com and sysadmins@example.com
> 6. send SMS notifications to NUMBER1 and only critical notifications to NUMBER2
> 7. execute your command /opt/company-scripts/restart-service.sh (keep in mind it will be run as user netdata)
> 8. etc.
that is brilliant I expect I can submit to the script parameters, but don't worry I will look into this myself already.

To extend the architecture I described, I have implemented Error Service (another micro - service in chain), so far in the moment it is responsible to distribute any kind of error to the target endpoints (either it is client web based app, email, support application backend, etc.). This error service is actually called Audit Service and it implements Audit pattern. It simply persists and do analysis on any kind of audit log (sequence of process flow from web BPM engine, every service call has ENTRY & EXIT logs). This service is not only about technical stuff, but greatly enhance the business side of view on any kind of event, so I can ultimately reconstruct any kind of scenario in other environment on click.

> I have tried to find some AI library to embed into netdata, so that it will continuously learn (without luck).

I am working on AI project, I experimented with Torch, TensorFlow, caffe2... but mainly work with deeplearning4j for our project needs, could be accelerated via Spark. I usually work on object/digits recognition, but I will discuss with folks in the wild on gitter and come back also with some ideas. The main fact is that we have available kind of object identification (label) and set of metrics (which is vector of key/value)... I will try to come with some idea. There exist techniques both requiring pretrained model, but also anomaly detection techniques without the model being available. Let's see.

**Comment by archenroot** [comment]
I also discuss this on: 
https://gitter.im/deeplearning4j/deeplearning4j

**Comment by archenroot** [comment]
No luck so far on gitter, I asked at: https://github.com/deeplearning4j/dl4j-examples/issues/577

**Comment by archenroot** [comment]
One interesting link I will be investigating:
https://www.ibm.com/developerworks/library/iot-deep-learning-anomaly-detection-3/

**Comment by archenroot** [comment]
And also: https://deeplearning4j.org/lstm.html, so LSTM networks in general

**Comment by ktsaou** [comment]
thanks! I would really love some help on AI... so, if you find something netdata can do, I would love to work on it...

**Comment by archenroot** [comment]
I'll get in touch, I have other AI tasks on my table :-) but I will analyze
and update here, lets leave this open for now.

2017-10-15 17:34 GMT+02:00 Costa Tsaousis <notifications@github.com>:

**Comment by ktsaou** [comment]
cleaning up. I am closing this. If you make some progress, post here and I'll be notified.
